backend/functions/src/emailing/functions.ts

```diff
@@ -45,9 +48,53 @@ export const updateEmailTimestamp = (
     .update({ [timestampField]: time })
 }
 
+export const updateIndivTimestamp = async (courseId: string, email: string) => {
+  const studentDocRef = studentRef.doc(email)
+  const studentDoc = await studentDocRef.get()
+
+  if (!studentDoc.exists) {
+    throw new Error(`Student document for ${email} does not exist`)
+  }
+  const studentData = studentDoc.data() as FirestoreStudent
+
+  const groups = studentData.groups
+  const groupMembership = groups.find((group) => group.courseId === courseId)
+  if (!groupMembership) {
+    throw new Error(`Student ${email} does not have membership in ${courseId}`)
+  }
+
+  groupMembership.firstNoMatchEmailTime = admin.firestore.Timestamp.now()
+
+  await studentDocRef.update({ groups })
+}
+
 /* ==== Emailing Helper Functions ==== */
 const GRAPH_ENDPOINT = 'https://graph.microsoft.com'
 
+export async function getRecipients(
+  courseId: string,
+  group: string,
+  indivEmail: string
+) {
+  let emailRcpts = ['lscstudypartners@cornell.edu']
+
+  if (parseInt(group) > 0) {
+    const groupData = await courseRef
+      .doc(courseId)
+      .collection('groups')
+      .doc(group)
+      .get()
+
+    emailRcpts = [...emailRcpts, ...(groupData.data() as any).members]
+  }
+
+  if (indivEmail !== undefined) {
+    emailRcpts = [...emailRcpts, indivEmail]
+  }
+
+  return emailRcpts
+}
```

Comment:
It feels weird to me that this function can still successfully get the recipients when the group and indivEmail combo is invalid. If both the group and individual email are defined it will return the individual email version.

Maybe create a function in `functions.ts` whose job is to get the recipients and send the email, and it does this by using `getRecipients` and `sendMails` (possibly `createEmailAsJson` too). Then the new function can be called in the `routes.ts` file to do everything, using these existing ones as helper functions.